Decrypt encrypted metadata

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->

Idea is to enable decryption of already encrypted metadata in decodeTransaction. The user specifies passphrase in which metadata was encrypted and the metadata located as in CIP83 is decrypted. The change is added in non-intrusive way. Due to lack of JSON instances exposure in cardano-api the needed functions were added (they are about to be erased when cardano-api exposes `metadataValueFromJsonNoSchema` in next node versions).

The heart of solution is `fromMetadataEncrypted`. Api spec was updated. Integration tests expanded to show the case.

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number
adp-3348

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
fromMetadataEncrypted - part 1

paweljakubas committed:
fromMetadataEncrypted - part 2

paweljakubas committed:
metadataValueFromJsonNoSchema

paweljakubas committed:
fromMetadataEncrypted - part 3

paweljakubas committed:
integrate fromMetadataEncryption and add IsSerrverError instance for ErrDecodetx

paweljakubas committed:
decodeTransaction integration tests - part 1

paweljakubas committed:
decodeTransaction integration tests - part 2

paweljakubas committed:
update api spec

paweljakubas committed:
fixes and adding missing bits to swagger

paweljakubas committed:
more fixes

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
````diff
@@ -116,7 +153,11 @@ Specifically:
+          }
     ...
     }
     ```
````
At the last line of this file (I can't link to it from this diff), could you remove the 4 spaces in front of `  Metadata encryption can be …`? They blow up the formatting.

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
```diff
@@ -6,11 +6,16 @@ In addition "Transactions New > Decode" HTTP endpoint is described in the contex
+input length must be a multiple of block size, ie., 16 bytes.
+PBKDF2 password stretching is used to get a 32-byte symmetric key
+that is required for the adopted encryption algorithm. In detail, 
+PBKDF2 encryption uses HMAC with the hash algorithm SHA512.
```
```suggestion
PBKDF2 encryption uses HMAC with the hash algorithm SHA512.

```

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
```diff
@@ -6,11 +6,16 @@ In addition "Transactions New > Decode" HTTP endpoint is described in the contex
 
-  [ref]: https://datatracker.ietf.org/doc/html/rfc7539
+  [ref1]: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0020
+  [ref2]: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0083
```
```suggestion
  [cip0020]: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0020
  [cip0083]: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0083
```

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
```diff
@@ -6,11 +6,16 @@ In addition "Transactions New > Decode" HTTP endpoint is described in the contex
+PBKDF2 password stretching is used to get a 32-byte symmetric key
+that is required for the adopted encryption algorithm. In detail, 
+PBKDF2 encryption uses HMAC with the hash algorithm SHA512.
 As a consequence the encrypted metadata, not its raw version, is going to be stored in blockchain.
```
```suggestion
As a consequence the encrypted metadata, not its raw version, is going to be stored in blockchain.
However, in line with [CIP-0020][cip0020] and [CIP-0083][cip0083], only the field `674` of the `metadata` field of the transaction will be affected.

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
```diff
@@ -6,11 +6,16 @@ In addition "Transactions New > Decode" HTTP endpoint is described in the contex
 Encryption of metadata is optional and when chosen the metadata in transaction is to be encrypted
-via AEAD scheme using ChaCha20 and Poly1305 (see [RFC 7539][ref]). PBKDF2 password stretching is used to get a 32-byte symmetric key
-that is required for the adopted encryption algorithm. In detail, PBKDF2 encryption uses HMAC with the hash algorithm SHA512.
+via AES256CBC according to [CIP-0020][ref1] and [CIP-0083][ref2].
```
```suggestion
via AES256CBC according to [CIP-0020][cip0020] and [cip0083][ref2].
```

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
````diff
@@ -55,15 +73,25 @@ Specifically:
     }
     ```
-    will return
+    will return (for salt "yoDCYXKaVhA=")
````
Uh, where does that salt come from?? 🤔 Is this a default value?

paweljakubas replied:
this is just an example salt generated for anyone eager to verify its encryption/decryption results

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -3264,6 +3280,104 @@ toMetadataEncrypted apiEncrypt payload saltM =
       where
         TxMetadata themap = payload ^. #txMetadataWithSchema_metadata
 
+-- When encryption is enabled we do the following:
```
```suggestion
-- When decryption is enabled we do the following:
```

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -3264,6 +3280,104 @@ toMetadataEncrypted apiEncrypt payload saltM =
+-- (c) decrypt each payload
+-- (d) update structure
+-- (e) decode metadata
+fromMetadataEncrypted
```
Good — but not in this module. 😅

Can you separate concerns better and split off a new module `Cardano.Wallet.Primitive.Types.Tx.TxMetadata` in the `/primitive` directory which contains all of these:

* `fromMetadataEncrypted`, `toMetadataEncrypted`, `cip83*`, `cip20*`, `metadataPBKDF2Config`

I believe that this move is only possible if you do not depend on the `ApiEncryptMetadata` type — which means that you need to use a different parameter. I suggest using a plain `ByteString` for the password, or defining a separate `Password` newtype in the new module.

paweljakubas replied:
I tried to relocate it to `primitive` but metadata is needed ... so in the end located it in `Cardano.Wallet.Api.Types.SchemaMetadata` module.

HeinrichApfelmus replied:
The JSON-related stuff should live in `Cardano.Wallet.Api.Types.SchemaMetadata`, that's alright

But: The encryption/decryption part does not depend on any JSON encoding (see comments), and I would like this independence to become manifest by moving the encryption functionality to a separate module in the `/primitive` directory as suggested. 🤓

paweljakubas replied:
relocated the stuff to ` Cardano.Wallet.Primitive.Types.MetadataEncryption` Also added extensive property tests and golden in ` Cardano.Wallet.Primitive.Types.MetadataEncryptionSpec`. Plus incorporated the condition that value in "msg" needs to be list of txts

HeinrichApfelmus commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -3765,6 +3766,17 @@ data ErrGetPolicyId
+    = ErrDecodeTxMissingMetadataKey
+    | ErrDecodeTxMissingEncryptionMethod
+    | ErrDecodeTxMissingValidEncryptionPayload
+    | ErrDecodeTxDecryptedPayload Text
```
```suggestion
    | ErrDecodeTxDecodePayload Text
```

The name of this error message has too much overlap with `ErrDecodeTxDecryptPayload `. Can you find names that better align with the message description? (The common prefix `ErrDecodeTx` is fine.)

paweljakubas replied:
oh I need to think here more - naming is not my strongest skills :sweat_smile:

HeinrichApfelmus replied:
How about these names, then:

```
ErrDecodeTxDecryptedPayload → ErrDecodeTxDecodingDecryptedPayloadFailed
ErrDecodeTxDecryptPayload → ErrDecodeTxDecryptionFailed
```

paweljakubas replied:
redefined errors in ` Cardano.Wallet.Primitive.Types.MetadataEncryption` separately for encryption and decryption.

HeinrichApfelmus reviewed (commented):
Thank you! First round of comments. 🤓

paweljakubas committed:
reformat metadata-encrypt.md

more

paweljakubas committed:
relocate metadata encryption - part 1

paweljakubas committed:
relocate metadata encryption - part 2

paweljakubas committed:
relocate metadata encryption - part 3

paweljakubas committed:
relocate metadata encryption - part 4

paweljakubas committed:
code standards

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
````diff
@@ -112,13 +151,17 @@ Specifically:
     ```
 
-  As a result we get decoded transaction with metadata decrypted:
+As a result we get decoded transaction with metadata decrypted:
````
```suggestion
As a result we get decoded transaction with metadata decrypted:

```

Could you add a newline to ensure that the code block is rendered correctly?

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
````diff
@@ -19,84 +27,115 @@ Specifically:
     }
     ```
-    will return
+will return (for the example salt "XG1cgIw56q8=")
````
```suggestion
will return (for the example salt "XG1cgIw56q8=")

```

Could you add a newline to ensure that the code block is rendered correctly?

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
````diff
@@ -19,84 +27,115 @@ Specifically:
     ```
 
-    Example:
+Example:
````
```suggestion
Example:

```

Could you add a newline to ensure that the code block is rendered correctly?

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
````diff
@@ -19,84 +27,115 @@ Specifically:
     }
     ```
-    will return
+will return (for the example salt "yoDCYXKaVhA=")
````
```suggestion
will return (for the example salt "yoDCYXKaVhA=")

```

Could you add a newline to ensure that the code block is rendered correctly?

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
```diff
@@ -19,84 +27,115 @@ Specifically:
+Metadata encryption can be used for shared wallet style when calling `/shared-wallets/{walletId}/transactions-construct` endpoint with the same `POST` payload.
 
-    Example:
+Example:
```
```suggestion
Example:

```

Could you add a newline to ensure that the code block is rendered correctly?

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
````diff
@@ -19,84 +27,115 @@ Specifically:
     ```
 
-    As a result we get transaction with metadata encrypted:
+As a result we get transaction with metadata encrypted:
````
```suggestion
As a result we get transaction with metadata encrypted:

```

Could you add a newline to ensure that the code block is rendered correctly?

HeinrichApfelmus commented on specifications/api/metadata-encrypt.md:
```diff
@@ -19,84 +27,115 @@ Specifically:
+In the `encrypt_metadata` field, passphrase used in encryption is established. `metadata` field to be encrypted is required.
 
-    Example `POST` data for the endpoint, ie.,   /wallets/{walletId}/transactions-construct`:
+Example `POST` data for the endpoint, ie.,   /wallets/{walletId}/transactions-construct`:
```
```suggestion
Example `POST` data for the endpoint, ie.,   /wallets/{walletId}/transactions-construct`:

```

Could you add a newline to ensure that the code block is rendered correctly?

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -99,3 +172,322 @@ instance FromJSON TxMetadataWithSchema where
+        | otherwise =
+            Left ErrConstructTxIncorrectRawMetadata
+      where
+        TxMetadata themap = payload ^. #txMetadataWithSchema_metadata
```
```suggestion
        TxMetadata themap = payload
```

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -99,3 +172,322 @@ instance FromJSON TxMetadataWithSchema where
+-- (d) add `enc` field with encryption method value 'basic'
+toMetadataEncrypted
+    :: ByteString
+    -> TxMetadataWithSchema
```
```suggestion
    -> TxMetadata
```

The encryption does not care about JSON schemas — by reflecting this in the types, we can move it out of the `Cardano.Wallet.Api` modules.

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -99,3 +172,322 @@ instance FromJSON TxMetadataWithSchema where
+--     [TxMetaText base64_1, TxMetaText base64_2, ..., TxMetaText base64_n]
+-- (d) add `enc` field with encryption method value 'basic'
+toMetadataEncrypted
+    :: ByteString
```
```suggestion
    :: Passphrase "lenient"
```

This type documents the purpose of the parameter more clearly.

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -99,3 +172,322 @@ instance FromJSON TxMetadataWithSchema where
+    updateTxMetadata :: TxMetadataValue -> TxMetadata
+    updateTxMetadata v = TxMetadata (Map.insert cip20MetadataKey v themap)
+      where
+        TxMetadata themap = payload ^. #txMetadataWithSchema_metadata
```
```suggestion
        TxMetadata themap = payload
```

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -99,3 +172,322 @@ instance FromJSON TxMetadataWithSchema where
+                Left ErrDecodeTxMissingValidEncryptionPayload
+            Right extracted
+
+    decodeFromJSON = ---use metadataValueFromJsonNoSchema when available from cardano-api
```
```suggestion
    decodeFromArray =
```

As mentioned above: We do not need to know about arbitrary JSON decodings, we only need to know how to decode an array of metadata strings from a string. This function has nothing to do with JSON.

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -99,3 +172,322 @@ instance FromJSON TxMetadataWithSchema where
+-- (d) update structure
+-- (e) decode metadata
+fromMetadataEncrypted
+    :: ByteString
```
```suggestion
    :: Passphrase "lenient"
```

HeinrichApfelmus commented on lib/api/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -99,3 +172,322 @@ instance FromJSON TxMetadataWithSchema where
+    Ledger.Map [ (toShelleyMetadatum k,
+                      toShelleyMetadatum v)
+                | (k,v) <- xs ]
+
```
I insist: Everything below this point should go into a module

`Cardano.Wallet.Primitive.Types.Tx.TxMetadata.Encryption`

in the `/primitive` directory. 🤓 By adjusting the types as I suggest, this will become possible. This work is necessary to achieve a good separation of concerns.

HeinrichApfelmus reviewed (commented):
Second round of review. I insist on the separation of concerns. 🤓 

After addressing the separation of concerns, I would also like to see property tests for

```
toMetadataEncrypted pwd . fromMetadataEncrypted pwd = id
fromMetadataEncrypted pwd . toMetadataEncrypted pwd = id
```

These two properties are vital to the whole feature and we need to cover them with extensive property tests in a way that integration tests cannot achieve.

paweljakubas committed:
relocation to primitive 1

relocation to primitive 2

relocation to primitive 3

relocation to primitive 4

relocation to primitive 5

paweljakubas committed:
add MetadataEncryptionSpec

paweljakubas committed:
extend golden

paweljakubas committed:
add roundtrip property

paweljakubas committed:
another property testes

paweljakubas committed:
make msg value obey being list of txts

adjust unit tests

another property

paweljakubas committed:
adjust integration testing

paweljakubas committed:
code standards

hlint

rebase cleanup

HeinrichApfelmus commented on lib/primitive/test/spec/Cardano/Wallet/Primitive/Types/MetadataEncryptionSpec.hs:
```diff
@@ -0,0 +1,490 @@
+            let encrypted = toMetadataEncrypted pwd' payload' (Just salt')
+            isRight encrypted ==>
+                (fromMetadataEncrypted pwd' (fromRight metadataNotValid encrypted))
+                === Right payload'
```
```suggestion
                (fromMetadataEncrypted pwd' =<< encrypted)
                === Right payload'
```

We don't expect encryption of the generated data to fail, do we? 🤔 In this case, using the fact that `Either e` is a monad helps.

If we do expect `encrypted` to match `Left _`, we would have to add a condition on the test coverage (using e.g. `cover`), otherwise we risk that the test succeeds vacuously (because `isRight encrypted = False`).

paweljakubas replied:
done

HeinrichApfelmus commented on lib/primitive/test/spec/Cardano/Wallet/Primitive/Types/MetadataEncryptionSpec.hs:
```diff
@@ -0,0 +1,490 @@
+            \(TestingSetup payload' pwd1 pwd2 salt') -> do
+            let encrypted = toMetadataEncrypted pwd1 payload' (Just salt')
+            isRight encrypted ==>
+                fromMetadataEncrypted pwd2 (fromRight metadataNotValid encrypted)
```
```suggestion
            fromMetadataEncrypted pwd2 =<< encrypted
```

See comment on `isRight encrypted` above.

paweljakubas replied:
done

HeinrichApfelmus commented on lib/primitive/test/spec/Cardano/Wallet/Primitive/Types/MetadataEncryptionSpec.hs:
```diff
@@ -0,0 +1,490 @@
+                        _ -> False
+            isRight encrypted ==>
+                fromRight metadataNotValid encrypted
+                `shouldSatisfy` hasCharacteristics
```
```suggestion
            (hasCharacteristics <$> encrypted)
                `shouldBe` Right True
```

See comment on `isRight encrypted` above.

paweljakubas replied:
done

HeinrichApfelmus commented on lib/primitive/test/spec/Cardano/Wallet/Primitive/Types/MetadataEncryptionSpec.hs:
```diff
@@ -0,0 +1,490 @@
+          )
+        ]
+      )
+    ]
```
```suggestion
```

The definition `metadataNotValid` can be removed if we don't use `fromRight` and `isRight encrypted` above.

paweljakubas replied:
done

HeinrichApfelmus commented on lib/primitive/lib/Cardano/Wallet/Primitive/Types/MetadataEncryption.hs:
```diff
@@ -0,0 +1,436 @@
+            _ -> []
+    extractTxt (TxMetaText txt) = txt
+    extractTxt _ =
+        error "TxMetaText is expected"
```
🤔 I have the feeling this call to `error` will hit when given malformed metadata of this kind:

```hs
example = Cardano.TxMetadata $
Map.fromList
[ ( 674
  , Cardano.TxMetaMap
    [ ( Cardano.TxMetaText "enc"
      , Cardano.TxMetaText "basic"
      )
    , ( Cardano.TxMetaText "msg"
      , Cardano.TxMetaList
        [ Cardano.TxMetaText "Fm/+xoZBA24yp8Vz548NAg=="
        , Cardano.TxMetaNumber 123 -- ouch
        ]
      )
    ]
  )
]
```

Can you convert this `error` into `Left Err…` instead?

In order to keep the code transparent, I recommend using `Maybe` as a monad. (You can also use `Either e`, but that is more cumbersome as you need to convert error types which is not worth doing here.) This technique often applies when you have many nested pattern matches — the combination of "validate the data with `filter` and then do the match" usually doesn't scale very well. (You seem to use this combination often, I recommend to not to that, but instead get accustomed to "pattern match as validation" instead.)

Here is one way of how you can do it:

```hs
getEncryptedPayload (TxMetaMap list) = do
    (_, TxMetaList xs) <- find isCip83EncryptPayload list
    txts <- for xs matchTxMetaText
    pure $ foldl' T.append T.empty txs
getEncryptedPayload _ =
    Nothing

matchTxMetaText (TxMetaText t) = Just t
matchTxMetaText _ = Nothing

isCip83EncryptPayload (TxMetaText k, TxMetaList _) =
    k == cip83EncryptPayloadKey
isCip83EncryptPayload _ =
    False
```

Note that I like repeating the function name `isCip83EncryptPayload` for each case — this is intentional, and very useful for equational reasoning; for example, this definition can also be read *in reverse*.

paweljakubas replied:
good idea, thanks. added handling of this case plus unit test checking this

HeinrichApfelmus reviewed (commented):
Much better, thank you! 😊 The new module is much more, well, modular, and also captures the possible error conditions nicely.

But I still have a couple of comments that I would like to see fixed before merging:
* `isRight encrypted` in the tests — having this here decreases my confidence in the tests unless we also use `cover`, fortunately, I don't think that it's necessary here.
* `error` in `fromMetadaEncrypted` — unfortunately, I think that this case can be hit when the metadata is sufficiently malformed. We need to turn this into a `Left` instead.

See detailed comments.

paweljakubas committed:
improve on unit tests - part 1

improve on unit tests - part 2

improve on unit tests - part 3 and add better error handling

paweljakubas committed:
code polishing

fix

HeinrichApfelmus reviewed (approved):
Thank you! 😊 Good to merge now — but could you squash the commits a little bit?
